flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/PreciseTimestamp.java

```diff
@@ -0,0 +1,234 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.dataformat;
+
+import org.apache.flink.table.types.logical.TimestampType;
+import org.apache.flink.util.Preconditions;
+
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+
+import static org.apache.calcite.avatica.util.DateTimeUtils.timestampStringToUnixDate;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestamp;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestampToString;
+
+/**
+ * Immutable SQL TIMESTAMP type with nanosecond precision.
+ */
+public class PreciseTimestamp implements Comparable<PreciseTimestamp> {
+
+	public static final int MAX_PRECISION = TimestampType.MAX_PRECISION;
+
+	public static final int MIN_PRECISION = TimestampType.MIN_PRECISION;
+
+	private final long milliseconds;
+
+	private final int nanoseconds;
+
+	private final int precision;
+
+	private PreciseTimestamp(long milliseconds, int nanoseconds, int precision) {
+		Preconditions.checkArgument(precision >= MIN_PRECISION && precision <= MAX_PRECISION);
+		Preconditions.checkArgument(nanoseconds >= 0 && nanoseconds <= 999_999_999);
+		this.milliseconds = milliseconds;
+		this.nanoseconds = nanoseconds;
+		this.precision = precision;
+	}
+
+	@Override
+	public int compareTo(PreciseTimestamp that) {
+		return this.milliseconds == that.milliseconds ?
+			Integer.compare(this.nanoseconds, that.nanoseconds) :
+			Long.compare(this.milliseconds, that.milliseconds);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (!(obj instanceof PreciseTimestamp)) {
+			return false;
+		}
+		PreciseTimestamp that = (PreciseTimestamp) obj;
+		return this.compareTo(that) == 0;
+	}
+
+	@Override
+	public String toString() {
+		// yyyy-MM-dd HH:mm:ss
+		StringBuilder timestampString = new StringBuilder(unixTimestampToString(milliseconds));
+		if (precision > 0 && nanoseconds > 0) {
+			// append fraction part
+			timestampString.append('.').append(nonosSecondsToString(nanoseconds, precision));
+		}
+		return timestampString.toString();
+	}
+
+	public long toLong() {
+		return milliseconds;
+	}
+
+	public static PreciseTimestamp fromLong(long milliseconds, int precision) {
+		Timestamp t = new Timestamp(milliseconds);
+		long millis = isCompact(precision) ?
+			zeroLastNDigits(t.getTime(), 3 - precision) : t.getTime();
+		int nanos = (int) zeroLastNDigits((long) t.getNanos(), 9 - precision);
+		return new PreciseTimestamp(millis, nanos, precision);
+	}
+
+	public Timestamp toTimestamp() {
```

Comment:
Should not have `java.sql.Timestamp` in this class.
This conversion should not depend on toString too.
Actually, `Timestamp` has `valueOf(LocalDateTime dateTime)` method.